modules/core/src/main/java/org/apache/ignite/internal/cdc/IgniteCDC.java

```diff
@@ -0,0 +1,485 @@
+ * @see DataStorageConfiguration#setCdcPath(String)
+ * @see DataStorageConfiguration#setWalForceArchiveTimeout(long)
+ * @see CommandLineStartup
+ * @see CaptureDataChangeConsumer
+ * @see DataStorageConfiguration#DFLT_CDC_PATH
+ */
+public class IgniteCDC implements Runnable {
+    /** State dir. */
+    public static final String STATE_DIR = "state";
+
+    /** Ignite configuration. */
+    private final IgniteConfiguration cfg;
+
+    /** CDC configuration. */
+    private final CaptureDataChangeConfiguration cdcCfg;
+
+    /** WAL iterator factory. */
+    private final IgniteWalIteratorFactory factory;
+
+    /** Events consumer. */
+    private final WALRecordsConsumer<?, ?> consumer;
+
+    /** Logger. */
+    private final IgniteLogger log;
+
+    /** CDC directory. */
+    private Path cdcDir;
+
+    /** Binary meta directory. */
+    private File binaryMeta;
+
+    /** Marshaller directory. */
+    private File marshaller;
+
+    /** CDC state. */
+    private CDCConsumerState state;
+
+    /** Save state to start from. */
+    private WALPointer initState;
+
+    /** Consistent ID. */
+    private final String nodeDir;
+
+    /** Previous segments. */
+    private final List<Path> prevSegments = new ArrayList<>();
+
+    /**
+     * @param cfg Ignite configuration.
+     * @param cdcCfg CDC configuration.
+     */
+    public IgniteCDC(IgniteConfiguration cfg, CaptureDataChangeConfiguration cdcCfg) {
+        this.cfg = cfg;
+        this.cdcCfg = cdcCfg;
+        this.consumer = new WALRecordsConsumer<>(cdcCfg.getConsumer());
+
+        log = logger(cfg, workDir(cfg));
+        factory = new IgniteWalIteratorFactory(log);
+
+        if (!CU.isPersistenceEnabled(cfg)) {
+            log.error("Persistence disabled. IgniteCDC can't run!");
+
+            throw new IllegalArgumentException("Persistence disabled. IgniteCDC can't run!");
+        }
+
+        nodeDir = consistentId(cfg);
+
+        if (nodeDir == null) {
+            log.warning("Can't determine nodeDir. It is recommended to set Consistent ID for production " +
+                "clusters(use IgniteConfiguration.setConsistentId or CaptureDataChangeConfiguration.setConsistentId " +
+                "and CaptureDataChangeConfiguration.setNodeIndex).");
+        }
+    }
+
+    /** Runs CDC. */
+    @Override public void run() {
+        try {
+            runX();
+        }
+        catch (Throwable e) {
+            e.printStackTrace();
+
+            throw new RuntimeException(e);
+        }
+    }
+
+    /** Runs CDC application with possible exception. */
+    public void runX() throws Exception {
+        if (log.isInfoEnabled()) {
+            log.info("Starting Ignite CDC Application.");
+            log.info("Consumer    -\t" + consumer.toString());
+        }
+
+        cdcDir = findCDCDir(workDir(cfg));
+
+        try (CDCFileLockHolder lock =
+                new CDCFileLockHolder(cdcDir.toString(), () -> "cdc.lock", log)) {
```

Comment:
I think the `id` of running consumer must be used instead of the `cdc.lock` name since multiple cdc-applications shares the same directory. In contrast, the Ignite instances use a single lock-file per directory.

Reply:
> use the id of running consumer instance instead of cdc-lock

Right now we don't have such a thing as "id", because only one concurrent consumer supported. :)